[ToolsVersion][Workspace] Abstract out the generation of Swift tools version specification

WowbaggersLiquidLunch opened the pull request:
Refactored `ManifestSourceGeneration.swift`, `Workspace.writeToolsVersion(at:version:fs:)`, and the latter's various callers, to extract the machine-generation of the Swift tools version specification into a new function `ToolsVersion.specification(resolution:)`.

### Motivation:

This was [very briefly discussed](https://github.com/apple/swift-package-manager/pull/3020#discussion_r519220009) in #3020. 

Bringing all the machine-generation of the tools version specification into a single function reduces clutter and boilerplate.

### Modifications:

- Introduced a new instance method of `ToolsVersion`: `specification(resolution:)`, and a helper type `ToolsVersion.SpecifierResolution`.

- Renamed `ToolsVersionWriter.swift` as `ToolsVersionSpecificationPrepender.swift`. The test file, class, and functions are also renamed accordingly.

- Renamed `Workspace.writeToolsVersion(at:version:fs:)` as `Workspace.prependToolsVersionSpecification(toDefaultManifestIn:specifying:fileSystem:)`, and delegates the main functionality to a new global function `Workspace.prependToolsVersionSpecification(toManifestAt:specifying:fileSystem:)`. The new function can write to version-specific manifests, in addition to `Package.swift`.

- Refactored `Manifest.generatedManifestFileContents`, `Workspace.prependToolsVersionSpecification(toDefaultManifestIn:specifying:fileSystem:)`, and the latter's various callers, to use the new function `ToolsVersion.specification(resolution:)`.

- Added more documentation, and some quality-of-life changes to make things follow the Swift API design guidelines more closely.

### Result:

As of commit 494b2b98a4390ca747b80c8e18ba8d7609ffd75b in this pull request, there is a possibility of a runtime crash from a `precondition`. I intend to change it to throw an error instead in this PR, but need advice on the best way to do it.

WowbaggersLiquidLunch commented on Sources/Workspace/ToolsVersionSpecificationPrepender.swift:
```diff
@@ -0,0 +1,92 @@
+///
+/// - Throws: A `FileSystemError` instance, if the manifest file is unable to be read from or written to..
+public func prependToolsVersionSpecification(toManifestAt manifestFilePath: AbsolutePath, specifying toolsVersion: ToolsVersion, fileSystem: FileSystem) throws {
+    // FIXME: Throw a `FileSystemError` instead?
+    // The only problem is that there doesn't seem to be a `FileSystemError.Kind` case that describes this kind of error.
+    // Or, we can revert it back to an assert, and let `fileSystem.readFileContents(manifestFilePath)` throw an error if the file can't be found.
+    precondition(fileSystem.isFile(manifestFilePath), "cannot locate the manifest file at \(manifestFilePath)")
```
I would like to draw attention to this FIXME. 

I changed this from an `assert` to a `precondition`. So now SwiftPM could potentially crash if the manifest file is not found at the path given. It's probably better if an error is thrown instead, but I can't find the correct error to throw in `TSCBasic.FileSystemError`. 

I could also change it back to an `assert`, which is ignored when SwiftPM is compiled for release, and the call to `FileSystem.readFileContents` 2 lines later throws an error if the manifest can't be found. The problem with this is that the error thrown (`FileSystemError.Kind.ioError` or `.noEntry` or `.isDirectory`) doesn't reflect the actual error.

abertelrud replied:
I'll take a look — this should definitely be an Error a not an assertion or precondition.

abertelrud replied:
I'll look at the existing code and the changes in the next few days and come up with a recommendation.

abertelrud replied:
I think that in this case, `noEntry` is probably the right kind of FileSystemError.  It should definitely not be an assertion or precondition.  The caller should be able to catch this error and report it.

I agree that this is not very specific, so the other, and maybe better, thing to do would be to define a specialization of Swift.Error (called for example ToolsVersionPresented.Error) similar to ToolsVersionLoader.Error that represents the semantic problem (no manifest) and then wraps the underlying error.

WowbaggersLiquidLunch replied:
I think I'll go with the latter option, so the user can receive a more specific feedback. Thanks for the suggestion!

WowbaggersLiquidLunch commented on Sources/Workspace/ToolsVersionSpecificationPrepender.swift:
```diff
@@ -0,0 +1,92 @@
+///   - fileSystem: The filesystem to read/write the manifest file on.
+///
+/// - Throws: A `FileSystemError` instance, if the manifest file is unable to be read from or written to..
+public func prependToolsVersionSpecification(toManifestAt manifestFilePath: AbsolutePath, specifying toolsVersion: ToolsVersion, fileSystem: FileSystem) throws {
```
This function is more general than `prependToolsVersionSpecification(toDefaultManifestIn:specifying:fileSystem:)`. It is able to write to version-specific manifest files. Should the function verify that the version to specify is no greater than the version in the manifest's file name? For example, reject using `6.0` as the version specifier if the manifest file is `Package@swift-5.0.swift`?

abertelrud replied:
Good question.  It kind of feels to me as if this function does a bit too much by generating the content as well as modifying the file in the file system.  Updating the tools version usually also involves changes to the manifest, so it doesn't feel like something that can be done in isolation.

WowbaggersLiquidLunch replied:
Sorry for this late response. What you wrote makes sense. 

Do you think it would be better if I just remove this function and the ability to write to version-specific manifests? Or, would it be better if I further factor this function to give each piece less responsibility?

WowbaggersLiquidLunch commented on Sources/PackageModel/ToolsVersionSpecificationGeneration.swift:
```diff
@@ -0,0 +1,47 @@
+    /// Returns a Swift tools version specification specifying the version to the given precision.
+    /// - Parameter resolution: The precision to which the version specifier follows the version.
+    /// - Returns: A  Swift tools version specification specifying the version to the given precision.
+    public func specification(resolution: SpecifierResolution = .automatic) -> String {
```
The parameter should be named better than `resolution`, but I can't think of a better one.

WowbaggersLiquidLunch replied:
What do you think of `rounding`, `rounded`, or `significantThrough`?

I think making the enum more specific to the patch version works too, but it will need more changes when/if `ToolsVersion` starts to carry version metadata.

abertelrud commented:
Would you consider renaming the file to "ToolsVersionSpecificationRewriter.swift" rather than "ToolsVersionSpecificationPrepender.swift", since it actually replaces the previous one and doesn't just prepend?  Also, once the package manifest editor functionality becomes more complete, setting the tools version might be something that should be accessed via that API, even if the actual implementation might be more string-oriented since this is modifying a comment.  See https://github.com/apple/swift-package-manager/pull/3034.

abertelrud commented:
@swift-ci please smoke test

abertelrud reviewed (changes requested):
I'd suggest changing the assert to a custom kind of Swift.Error if the package manifest can't be found, and I'd suggest "rewriter" for the name-changed source file, but apart from that I think this looks pretty good.

WowbaggersLiquidLunch commented:
> Would you consider renaming the file to "ToolsVersionSpecificationRewriter.swift" rather than "ToolsVersionSpecificationPrepender.swift", since it actually replaces the previous one and doesn't just prepend?

Yep, I can rename it. It does prepend a new one in some situations, but yes I agree the term "prepend" isn't really accurate. 

Alternatively, what do you think of something like "ManifestReversioner"? I think it's more concise and highlights that it changes the manifest file.

WowbaggersLiquidLunch commented:
Sorry for the long delay since my previous update. 

Summary of new changes:

1. Inaccessible manifest now throws an error, instead of crashing SwiftPM. I think this is the most valuable change in this PR.

2. The word "prepend" in file and method names are replaced by "rewrite", to more accurately reflect what they do.

3. Removed from this PR the ability to change the tools version specification line in version-specific manifest. Now only the main manifest can be changed. Now the functionality is much the same as in the main branch.

4. Reworded the label and enum in `ToolsVersionSpecificationGeneration.swift`.

5. Rebased to main (ec407ac14738bf132b23441aa9435a919124eda6).

abertelrud commented:
Thanks for the changes!  This looks good to me.

abertelrud commented:
@swift-ci please smoke test

tomerd commented:
@abertelrud @WowbaggersLiquidLunch is this ready to merge?

WowbaggersLiquidLunch commented:
It's ready in my opinion.

neonichu commented:
Thanks for the PR @WowbaggersLiquidLunch and sorry for the delay on this. Would you mind resolving the conflicts, then we can re-run CI and get this merged.

tomerd commented:
@swift-ci please smoke test

WowbaggersLiquidLunch commented:
The rebase I did a few minutes ago resolves the merge conflict.

It seems that some of the commits can be squashed, but I don't remember much of my reasoning for those changes, so I just left them be.

WowbaggersLiquidLunch commented:
I forgot to test locally after the previous rebase, and just rebased again to fix some tests.

The previous merge conflict was (partially?) because `Manifest.generatedManifestFileContents` adds a space after the colon if Swift version ≥ 5.4, so I changed `ToolsVersion.specification` to do the same. However, there are other places than `Manifest.generatedManifestFileContents` that start to use `ToolsVersion.specification` in this PR, so I changed some tests to expect a space after the colon in the tools version specification for Swift ≥ 5.4.

All tests (except for those that skip on macOS) now pass on my computer.

neonichu commented:
@swift-ci please smoke test

abertelrud commented:
This looks good to me to merge at this point.  Thanks again @WowbaggersLiquidLunch for your work on this, and your patience!
